examples/pi-home-router/src/arp/arp_handler.rs

```diff
@@ -0,0 +1,212 @@
+use crate::types::InterfaceAnnotated;
+use route_rs_packets::{ArpFrame, ArpHardwareType, ArpOp, EthernetFrame, MacAddr, IPV4_ETHER_TYPE};
+use route_rs_runtime::processor::Processor;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::net::{IpAddr, Ipv4Addr};
+use std::sync::{Arc, Mutex};
+
+// TODO: We need to respond to ARP requests that are targeted at the router.
+// Where is this state stored? How to we access it?
+const ROUTER_IPV4_ADDR: Ipv4Addr = Ipv4Addr::LOCALHOST;
+
+pub(crate) struct ArpHandler {
+    arp_table: Arc<Mutex<HashMap<Ipv4Addr, MacAddr>>>,
+}
+
+impl ArpHandler {
+    pub fn new(arp_table: Arc<Mutex<HashMap<Ipv4Addr, MacAddr>>>) -> Self {
+        ArpHandler { arp_table }
+    }
+}
+
+impl Processor for ArpHandler {
+    type Input = InterfaceAnnotated<EthernetFrame>;
+    type Output = InterfaceAnnotated<EthernetFrame>;
+
+    ///
+    /// From the ARP RFC: https://tools.ietf.org/html/rfc826
+    /// Handler assumes it receives ARP packets
+    ///
+    /// ?Do I have the hardware type in ar$hrd?
+    /// Yes: (almost definitely)
+    ///     [optionally check the hardware length ar$hln]
+    ///     ?Do I speak the protocol in ar$pro?
+    ///     Yes:
+    ///         [optionally check the protocol length ar$pln]
+    ///         Merge_flag := false
+    ///         If the pair <protocol type, sender protocol address> is already in my translation
+    ///             table, update the sender hardware address field of the entry with the new
+    ///             information in the packet and set Merge_flag to true.
+    ///         ?Am I the target protocol address?
+    ///         Yes:
+    ///             If Merge_flag is false, add the triplet <protocol type, sender protocol address,
+    ///                 sender hardware address> to the translation table.
+    ///             ?Is the opcode ares_op$REQUEST?  (NOW look at the opcode!!)
+    ///             Yes:
+    ///                 Swap hardware and protocol fields, putting the local hardware and protocol
+    ///                     addresses in the sender fields.
+    ///                 Set the ar$op field to ares_op$REPLY
+    ///                 Send the packet to the (new) target hardware address on the same hardware on
+    ///                     which the request was received.
+    ///
+    fn process(&mut self, packet: Self::Input) -> Option<Self::Output> {
+        let mut arp_frame = ArpFrame::try_from(packet.packet).unwrap();
```

Comment:
There's a few different options for how we can handle invalid ARP frames. We can either
1. Do validation outside of this Processor so `Self::Input == ArpFrame` (a little tighter bounds)
2. Validate here and drop on error
3. Validate here, log error, then drop
4. Validate here, forward to logging element which drops.

Reply:
Can we use the type system here to provide us an IntAnt<ArpFrame>. Seems closest to what we wanted this system to do in the first place, use the static types to prevent errors.

Reply:
My second choice is to drop on error, ignore logging for now.